explanation/performance/perf-pgo.md

```diff
@@ -0,0 +1,226 @@
+
+```bash
+$ file /usr/bin/openssl
+/usr/bin/openssl: ELF 64-bit LSB pie executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=42a5ac797b6981bd45a9ece3f67646edded56bdb, for GNU/Linux 3.2.0, with debug_info, not stripped
+```
+
+Note how `file` reports the binary as being `with debug_info, not stripped`.  This is the output you need to look for when inspecting your own binaries.
+
+You can use the `eu-readelf` command (from the `elfutils` package) to determine whether any compression was performed with `dwz`:
+
+```bash
+$ eu-readelf -S /usr/bin/openssl
+[...]
+Section Headers:
+[Nr] Name                 Type         Addr             Off      Size     ES Flags Lk Inf Al
+[...]
+[31] .debug_abbrev        PROGBITS     0000000000000000 001bcbf3 000110c4  0        0   0  1
+[...]
+```
+
+If the `Flags` field has `C` (as in `compressed`) in it, this means that `dwz` was used.  You will then need to figure out how to disable its execution during the build; for this particular build, this was done by overriding the `dh_dwz` rule on `debian/rules`, setting it to an empty value, and executing the build again.  The particularities of how to skip build steps depend on the build system being used, and as such this guide will not go into further details.
+
+### Using `perf` to obtain profile data
+
+With OpenSSL prepared and ready to be profiled, it's now time to use `perf` to obtain profiling data from our workload.  First, let's make sure we can actually access profiling data in the system.  As `root`, run:
+
+```bash
+# echo -1 > /proc/sys/kernel/perf_event_paranoid
+```
+
+This should allow all users to monitor events in the system.
+
+Then, if you are on an Intel system, you can invoke `perf` using:
+
+```bash
+$ sudo perf record \
+	-e br_inst_retired.near_taken \
+	--branch-any \
+	--all-cpus \
+	--output openssl-nonpgo.perfdata \
+	-- openssl speed -mr -seconds 60 -evp md5 sha512 rsa2048 aes-256-cbc
+```
+
+On an AMD system, use the following invocation:
+
+```bash
+$ sudo perf record \
+	-e 'cpu/event=0xc4,umask=0x0,name=ex_ret_brn_tkn/' \
+	--branch-any \
+	--all-cpus \
+	--output openssl-nonpgo.perfdata \
+	-- openssl speed -mr -seconds 60 -evp md5 sha512 rsa2048 aes-256-cbc
+```
+
+After the command has finished running, you should see a file named `openssl-nonpgo.perfdata` in your current directory.
+
+Note that the only thing that differs between the Intel and AMD variants is the PMU event to be monitored.  Also, note how we are using `sudo` to invoke `perf record`.  This is necessary in order to obtain full access to Linux kernel symbols and relocation information.  It also means that the `openssl-nonpgo.perfdata` file ownership will need to be adjusted:
+
+```bash
+$ sudo chown ubuntu:ubuntu openssl-nonpgo.perfdata
+```
+
+Now, we need to convert the file to `gcov`.
+
+#### Alternative scenario: profiling the program using other approaches
```

Comment:
Looking at the rendered version, I'm a bit confused why this is so indented - since it's not really part of the first subsection, should it be promoted to a level 2 header? (##)

Reply:
Hm, the problem with promoting this to a level 2 header is that it breaks the flow of instructions.  Let me see if I can move this subsection down.